twilight-gateway/src/shard.rs

```diff
@@ -489,13 +493,29 @@ impl Shard {
     ///
     /// Returns a [`ReceiveMessageErrorType::SendingMessage`] error type if the
     /// shard failed to send a message to the gateway, such as a heartbeat.
+    ///
+    /// [`next_message`]: Self::next_message
     pub async fn next_event(&mut self) -> Result<Event, ReceiveMessageError> {
         loop {
-            match self.next_message().await? {
+            let text = match self.next_message().await? {
                 Message::Close(frame) => return Ok(Event::GatewayClose(frame)),
-                Message::Text(event) => {
-                    if let Some(event) = json::parse(event, self.config.event_types())? {
-                        return Ok(event.into());
+                Message::Text(text) => text,
+            };
+
+            match json::parse(text, self.config.event_types()) {
+                Ok(Some(event)) => return Ok(event.into()),
+                Ok(None) => {}
+                Err(source) => {
+                    // Discord has many events that aren't documented, so we
+                    // need to skip over errors caused by unknown events or
+                    // opcodes.
+                    #[allow(clippy::redundant_closure_for_method_calls)]
+                    if source
+                        .source()
+                        .and_then(|source| source.downcast_ref::<UnknownEventError>())
+                        .is_some()
+                    {
+                        continue;
```

Comment:
We should definately log here, however, at which level is unclear. Unknown opcodes, which are very rare, are already logged at `INFO` in `Shard::process` but the reason for logging differs. In `Shard::process` unknown opcodes are a problem due to the shard maybe not processing important information whereas here it's a problem due to not being able to return the event to the user. Unknown dispatch events account for most unknown event errors and their type is already logged at `DEBUG` in `Shard::process` so repeating their type might not be necessary.

Logging at `INFO` to nag users (to nag Twilight) about unknown events could be considered a "feature", but to not cause log spam them it might be better to log at `DEBUG`. Additionally, Discord advertises new dispatch events so nagging users may be redundant.

All in all we should atleast log that we skipped an event at `DEBUG` and users will know which type it is due to the context from the previous log in `Shard::process`. It's important that we include attach a span to the logged event, but we can't instrument the whole `next_event` method as that'd duplicate with `next_message`'s span.

```suggestion
                        let span = tracing::info_span!(id = %self.id, "shard");
                        tracing::debug!(parent: &span, "skipped deserializing unknown event");
                        continue;
```

Reply:
your suggestion doesnt compile

Reply:
Sorry, the ordering of the span's arguments should be reversed, `let span = tracing::info_span!("shard", id=%self.id);` should work.